File: EmployeeSchedulerGUI.java
/*
 * Java Class to implement employee scheduling mechanism.
 * @author Unique Karanjit 
 * 
 */

import javax.swing.*;
import java.awt.*;
import java.util.*;
import java.util.List;  // Explicitly import java.util.List

public class EmployeeSchedulerGUI {
    private JTextArea scheduleArea;
    private JButton printButton;
    private Set<String> employeeList;
    private CardLayout cardLayout;
    private JPanel cards;
    private EmployeeScheduler scheduler;
    private JComboBox<String> employeeComboBox;

    public EmployeeSchedulerGUI() {
        employeeList = new HashSet<>();
        scheduler = new EmployeeScheduler();
        employeeComboBox = new JComboBox<>();  // Initialize here
        createGUI();
    }

    private void createGUI() {
        JFrame frame = new JFrame("Employee Scheduler");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setSize(1200, 800);  // Increased window size

        JTabbedPane tabbedPane = new JTabbedPane();
        
        // Employee Management Panel
        tabbedPane.addTab("Employee Management", createEmployeePanel());
        
        // Schedule Management Panel
        tabbedPane.addTab("Schedule Management", createSchedulePanel());

        frame.add(tabbedPane);
        frame.setLocationRelativeTo(null);  // Center on screen
        frame.setVisible(true);
    }

    private JPanel createEmployeePanel() {
        JPanel panel = new JPanel(new BorderLayout(10, 10));  // Added spacing
        
        // North panel for buttons
        JPanel buttonPanel = new JPanel(new FlowLayout());
        JButton addButton = new JButton("Add Employee");
        JButton viewButton = new JButton("View Employees");
        buttonPanel.add(addButton);
        buttonPanel.add(viewButton);
        
        // Center panel for content
        JPanel contentPanel = new JPanel(new BorderLayout());
        cardLayout = new CardLayout();
        cards = new JPanel(cardLayout);
        
        // Add Employee Panel
        JPanel addPanel = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5, 5, 5, 5);
        gbc.anchor = GridBagConstraints.WEST;
        
        // Employee name field
        gbc.gridx = 0; gbc.gridy = 0;
        addPanel.add(new JLabel("Employee Name:"), gbc);
        gbc.gridx = 1;
        JTextField newEmployeeField = new JTextField(20);
        addPanel.add(newEmployeeField, gbc);

        // First preference
        gbc.gridx = 0; gbc.gridy = 1;
        addPanel.add(new JLabel("1st Preference Day:"), gbc);
        String[] days = {"None", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
        JComboBox<String> firstDayCombo = new JComboBox<>(days);
        gbc.gridx = 1;
        addPanel.add(firstDayCombo, gbc);

        gbc.gridx = 0; gbc.gridy = 2;
        addPanel.add(new JLabel("1st Preference Shift:"), gbc);
        String[] shifts = {"None", "Morning", "Afternoon", "Evening"};
        JComboBox<String> firstShiftCombo = new JComboBox<>(shifts);
        gbc.gridx = 1;
        addPanel.add(firstShiftCombo, gbc);

        // Second preference
        gbc.gridx = 0; gbc.gridy = 3;
        addPanel.add(new JLabel("2nd Preference Day:"), gbc);
        JComboBox<String> secondDayCombo = new JComboBox<>(days);
        gbc.gridx = 1;
        addPanel.add(secondDayCombo, gbc);

        gbc.gridx = 0; gbc.gridy = 4;
        addPanel.add(new JLabel("2nd Preference Shift:"), gbc);
        JComboBox<String> secondShiftCombo = new JComboBox<>(shifts);
        gbc.gridx = 1;
        addPanel.add(secondShiftCombo, gbc);

        // Third preference
        gbc.gridx = 0; gbc.gridy = 5;
        addPanel.add(new JLabel("3rd Preference Day:"), gbc);
        JComboBox<String> thirdDayCombo = new JComboBox<>(days);
        gbc.gridx = 1;
        addPanel.add(thirdDayCombo, gbc);

        gbc.gridx = 0; gbc.gridy = 6;
        addPanel.add(new JLabel("3rd Preference Shift:"), gbc);
        JComboBox<String> thirdShiftCombo = new JComboBox<>(shifts);
        gbc.gridx = 1;
        addPanel.add(thirdShiftCombo, gbc);

        // Submit button
        gbc.gridx = 1; gbc.gridy = 7;
        JButton submitButton = new JButton("Submit");
        addPanel.add(submitButton, gbc);
        
        // View Employees Panel
        JPanel viewPanel = new JPanel(new BorderLayout(10, 10));  // Added spacing
        JTextArea employeeListArea = new JTextArea(25, 50);  // Increased size
        employeeListArea.setEditable(false);
        employeeListArea.setFont(new Font("Monospaced", Font.PLAIN, 14));  // Larger font
        JScrollPane viewScrollPane = new JScrollPane(employeeListArea);
        viewPanel.add(viewScrollPane, BorderLayout.CENTER);
        
        cards.add(addPanel, "ADD");
        cards.add(viewPanel, "VIEW");
        contentPanel.add(cards, BorderLayout.CENTER);
        
        panel.add(buttonPanel, BorderLayout.NORTH);
        panel.add(contentPanel, BorderLayout.CENTER);

        // View Button Action
        viewButton.addActionListener(e -> {
            try {
                if (employeeList.isEmpty()) {
                    employeeListArea.setText("No employees found.");
                } else {
                    StringBuilder sb = new StringBuilder("Current Employees:\n\n");
                    for (String employee : employeeList) {
                        sb.append("• ").append(employee);
                        
                        Map<String, List<String>> employeePrefs = scheduler.getEmployeePreferences(employee);
                        if (employeePrefs != null && !employeePrefs.isEmpty()) {
                            sb.append("\n  Preferences:");
                            for (Map.Entry<String, List<String>> entry : employeePrefs.entrySet()) {
                                String day = entry.getKey();
                                ArrayList<String> shiftList = new ArrayList<>(entry.getValue());
                                sb.append("\n    ").append(day).append(": ")
                                  .append(String.join(", ", shiftList));
                            }
                        } else {
                            sb.append("\n  No preferences specified");
                        }
                        sb.append("\n\n");
                    }
                    employeeListArea.setText(sb.toString());
                }
                cardLayout.show(cards, "VIEW");
            } catch (Exception ex) {
                System.err.println("Error in view button: " + ex.getMessage());
                ex.printStackTrace();
                JOptionPane.showMessageDialog(panel, 
                    "Error displaying employees: " + ex.getMessage(),
                    "Error",
                    JOptionPane.ERROR_MESSAGE);
            }
        });

        // Add Button Action
        addButton.addActionListener(e -> {
            cardLayout.show(cards, "ADD");
            newEmployeeField.setText("");
            firstDayCombo.setSelectedItem("None");
            firstShiftCombo.setSelectedItem("None");
            secondDayCombo.setSelectedItem("None");
            secondShiftCombo.setSelectedItem("None");
            thirdDayCombo.setSelectedItem("None");
            thirdShiftCombo.setSelectedItem("None");
        });

        // Submit Button Action
        submitButton.addActionListener(e -> {
            try {
                String name = newEmployeeField.getText().trim();
                String firstDay = (String) firstDayCombo.getSelectedItem();
                String firstShift = (String) firstShiftCombo.getSelectedItem();
                String secondDay = (String) secondDayCombo.getSelectedItem();
                String secondShift = (String) secondShiftCombo.getSelectedItem();
                String thirdDay = (String) thirdDayCombo.getSelectedItem();
                String thirdShift = (String) thirdShiftCombo.getSelectedItem();

                if (name.isEmpty()) {
                    JOptionPane.showMessageDialog(panel, 
                        "Please enter an employee name.",
                        "Invalid Input",
                        JOptionPane.WARNING_MESSAGE);
                    return;
                }

                // Add employee to the list
                employeeList.add(name);
                
                // Always add employee to scheduler, even without preferences
                scheduler.addEmployee(name);

                // Add preferences only if they're not "None"
                if (!firstDay.equals("None") && !firstShift.equals("None")) {
                    scheduler.addEmployeePreference(name, firstDay, firstShift);
                }
                if (!secondDay.equals("None") && !secondShift.equals("None")) {
                    scheduler.addEmployeePreference(name, secondDay, secondShift);
                }
                if (!thirdDay.equals("None") && !thirdShift.equals("None")) {
                    scheduler.addEmployeePreference(name, thirdDay, thirdShift);
                }

                // Update UI components
                updateEmployeeComboBox();
                
                // Show success message with only the actual preferences
                StringBuilder message = new StringBuilder("Employee " + name + " added successfully");
                if (!firstDay.equals("None") && !firstShift.equals("None")) {
                    message.append("\n1st Choice: ").append(firstDay).append(" ").append(firstShift);
                }
                if (!secondDay.equals("None") && !secondShift.equals("None")) {
                    message.append("\n2nd Choice: ").append(secondDay).append(" ").append(secondShift);
                }
                if (!thirdDay.equals("None") && !thirdShift.equals("None")) {
                    message.append("\n3rd Choice: ").append(thirdDay).append(" ").append(thirdShift);
                }
                if (firstDay.equals("None") || firstShift.equals("None")) {
                    message.append("\n(No preferences specified)");
                }
                
                JOptionPane.showMessageDialog(panel, 
                    message.toString(),
                    "Success",
                    JOptionPane.INFORMATION_MESSAGE);

                // Clear input fields
                newEmployeeField.setText("");
                firstDayCombo.setSelectedItem("None");
                firstShiftCombo.setSelectedItem("None");
                secondDayCombo.setSelectedItem("None");
                secondShiftCombo.setSelectedItem("None");
                thirdDayCombo.setSelectedItem("None");
                thirdShiftCombo.setSelectedItem("None");

            } catch (Exception ex) {
                JOptionPane.showMessageDialog(panel,
                    "Error adding employee: " + ex.getMessage(),
                    "Error",
                    JOptionPane.ERROR_MESSAGE);
                ex.printStackTrace();
            }
        });

        return panel;
    }


    private void updateEmployeeComboBox() {
        employeeComboBox.removeAllItems();
        employeeComboBox.addItem("None");
        for (String employee : employeeList) {
            employeeComboBox.addItem(employee);
        }
    }

    private JPanel createSchedulePanel() {
        JPanel panel = new JPanel(new BorderLayout());  // Changed to BorderLayout
        
        // Button panel
        JPanel buttonPanel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        printButton = new JButton("Generate Schedule");
        buttonPanel.add(printButton);
        buttonPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));
        panel.add(buttonPanel, BorderLayout.NORTH);

        // Center panel to hold both text areas
        JPanel centerPanel = new JPanel(new BorderLayout(0, 10));  // 10 pixel vertical gap
        centerPanel.setBorder(BorderFactory.createEmptyBorder(0, 10, 10, 10));  // Add padding

        // Schedule display area
        scheduleArea = new JTextArea(30, 100);
        scheduleArea.setEditable(false);
        scheduleArea.setFont(new Font("Monospaced", Font.PLAIN, 14));
        JScrollPane scrollPane = new JScrollPane(scheduleArea);
        centerPanel.add(scrollPane, BorderLayout.CENTER);

        // Current preferences display area
        JTextArea preferencesArea = new JTextArea(10, 100);
        preferencesArea.setEditable(false);
        preferencesArea.setFont(new Font("Monospaced", Font.PLAIN, 14));
        JScrollPane prefScrollPane = new JScrollPane(preferencesArea);
        centerPanel.add(prefScrollPane, BorderLayout.SOUTH);

        panel.add(centerPanel, BorderLayout.CENTER);

        // Generate Schedule Button Action
        printButton.addActionListener(e -> {
            if (employeeList.size() < 3) {
                JOptionPane.showMessageDialog(panel,
                    "At least 3 employees are required to generate a schedule.\n" +
                    "Current number of employees: " + employeeList.size(),
                    "Insufficient Employees",
                    JOptionPane.WARNING_MESSAGE);
                return;
            }

            String result = scheduler.generateSchedule();
            if (result.startsWith("Error:") || result.startsWith("Warning:")) {
                JOptionPane.showMessageDialog(panel,
                    result,
                    "Schedule Generation Issue",
                    JOptionPane.WARNING_MESSAGE);
            } else {
                scheduleArea.setText(result);
                preferencesArea.setText(scheduler.getPreferencesText());
            }
        });

        return panel;
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(EmployeeSchedulerGUI::new);
    }
}

File: EmployeeSchedulerGUI.py
"""
Python Class to implement employee scheduling GUI.
@author Unique Karanjit 
"""
# EmployeeSchedulerGUI.py
import tkinter as tk
from tkinter import ttk, messagebox
from EmployeeScheduler import EmployeeScheduler

class EmployeeSchedulerGUI:
    """GUI for Employee Scheduler application"""
    
    def __init__(self):
        self.scheduler = EmployeeScheduler()
        self.employee_list = set()
        
        # Create main window
        self.root = tk.Tk()
        self.root.title("Employee Scheduler")
        self.root.geometry("1200x800")
        
        # Create notebook (tabbed interface)
        self.notebook = ttk.Notebook(self.root)
        self.notebook.pack(fill='both', expand=True, padx=10, pady=5)
        
        # Create tabs
        self.create_employee_tab()
        self.create_schedule_tab()
        
    def create_employee_tab(self):
        """Creates the Employee Management tab"""
        employee_frame = ttk.Frame(self.notebook)
        self.notebook.add(employee_frame, text="Employee Management")
        
        # Employee input section
        input_frame = ttk.LabelFrame(employee_frame, text="Add New Employee", padding=10)
        input_frame.pack(fill='x', padx=10, pady=5)
        
        # Name input
        ttk.Label(input_frame, text="Employee Name:").grid(row=0, column=0, sticky='w')
        self.name_entry = ttk.Entry(input_frame, width=30)
        self.name_entry.grid(row=0, column=1, padx=5, pady=5)
        
        # Preference inputs
        days = ["None"] + EmployeeScheduler.DAYS_OF_WEEK
        shifts = ["None"] + EmployeeScheduler.SHIFT_TIMES
        
        # First preference
        ttk.Label(input_frame, text="1st Choice:").grid(row=1, column=0, sticky='w')
        self.first_day = ttk.Combobox(input_frame, values=days, state="readonly")
        self.first_day.set("None")
        self.first_day.grid(row=1, column=1, padx=5, pady=5)
        self.first_shift = ttk.Combobox(input_frame, values=shifts, state="readonly")
        self.first_shift.set("None")
        self.first_shift.grid(row=1, column=2, padx=5, pady=5)
        
        # Second preference
        ttk.Label(input_frame, text="2nd Choice:").grid(row=2, column=0, sticky='w')
        self.second_day = ttk.Combobox(input_frame, values=days, state="readonly")
        self.second_day.set("None")
        self.second_day.grid(row=2, column=1, padx=5, pady=5)
        self.second_shift = ttk.Combobox(input_frame, values=shifts, state="readonly")
        self.second_shift.set("None")
        self.second_shift.grid(row=2, column=2, padx=5, pady=5)
        
        # Third preference
        ttk.Label(input_frame, text="3rd Choice:").grid(row=3, column=0, sticky='w')
        self.third_day = ttk.Combobox(input_frame, values=days, state="readonly")
        self.third_day.set("None")
        self.third_day.grid(row=3, column=1, padx=5, pady=5)
        self.third_shift = ttk.Combobox(input_frame, values=shifts, state="readonly")
        self.third_shift.set("None")
        self.third_shift.grid(row=3, column=2, padx=5, pady=5)
        
        # Submit button
        submit_btn = ttk.Button(input_frame, text="Add Employee", command=self.add_employee)
        submit_btn.grid(row=4, column=0, columnspan=3, pady=10)
        
        # Employee list display
        list_frame = ttk.LabelFrame(employee_frame, text="Current Employees", padding=10)
        list_frame.pack(fill='both', expand=True, padx=10, pady=5)
        
        self.employee_text = tk.Text(list_frame, height=20, width=70)
        self.employee_text.pack(fill='both', expand=True)
        
    def create_schedule_tab(self):
        """Creates the Schedule Management tab"""
        schedule_frame = ttk.Frame(self.notebook)
        self.notebook.add(schedule_frame, text="Schedule Management")
        
        # Generate button
        generate_btn = ttk.Button(schedule_frame, text="Generate Schedule", 
                                command=self.generate_schedule)
        generate_btn.pack(pady=10)
        
        # Schedule display
        self.schedule_text = tk.Text(schedule_frame, height=30, width=100)
        self.schedule_text.pack(fill='both', expand=True, padx=10, pady=5)
        
    def add_employee(self):
        """Handles adding a new employee"""
        try:
            name = self.name_entry.get().strip()
            if not name:
                messagebox.showwarning("Invalid Input", "Please enter an employee name.")
                return
            
            # Add to employee list
            self.employee_list.add(name)
            
            # Add to scheduler
            self.scheduler.add_employee(name)
            
            # Add preferences if specified
            preferences_added = []
            
            if self.first_day.get() != "None" and self.first_shift.get() != "None":
                self.scheduler.add_employee_preference(name, self.first_day.get(), 
                                                    self.first_shift.get())
                preferences_added.append(f"1st Choice: {self.first_day.get()} "
                                      f"{self.first_shift.get()}")
            
            if self.second_day.get() != "None" and self.second_shift.get() != "None":
                self.scheduler.add_employee_preference(name, self.second_day.get(), 
                                                    self.second_shift.get())
                preferences_added.append(f"2nd Choice: {self.second_day.get()} "
                                      f"{self.second_shift.get()}")
            
            if self.third_day.get() != "None" and self.third_shift.get() != "None":
                self.scheduler.add_employee_preference(name, self.third_day.get(), 
                                                    self.third_shift.get())
                preferences_added.append(f"3rd Choice: {self.third_day.get()} "
                                      f"{self.third_shift.get()}")
            
            # Show success message
            message = f"Employee {name} added successfully"
            if preferences_added:
                message += "\n" + "\n".join(preferences_added)
            else:
                message += "\n(No preferences specified)"
            
            messagebox.showinfo("Success", message)
            
            # Update employee list display
            self.update_employee_list()
            
            # Clear inputs
            self.clear_inputs()
            
        except Exception as e:
            messagebox.showerror("Error", f"Error adding employee: {str(e)}")
            
    def generate_schedule(self):
        """Handles schedule generation"""
        try:
            result = self.scheduler.generate_schedule()
            self.schedule_text.delete(1.0, tk.END)
            self.schedule_text.insert(tk.END, result)
        except Exception as e:
            messagebox.showerror("Error", f"Error generating schedule: {str(e)}")
            
    def update_employee_list(self):
        """Updates the employee list display"""
        self.employee_text.delete(1.0, tk.END)
        if not self.employee_list:
            self.employee_text.insert(tk.END, "No employees found.")
            return
        
        text = "Current Employees:\n\n"
        for employee in sorted(self.employee_list):
            text += f"• {employee}\n"
            prefs = self.scheduler.preferences.get(employee, {})
            if prefs:
                text += "  Preferences:\n"
                for day, shifts in prefs.items():
                    text += f"    {day}: {', '.join(shifts)}\n"
            else:
                text += "  No preferences specified\n"
            text += "\n"
        
        self.employee_text.insert(tk.END, text)
        
    def clear_inputs(self):
        """Clears all input fields"""
        self.name_entry.delete(0, tk.END)
        self.first_day.set("None")
        self.first_shift.set("None")
        self.second_day.set("None")
        self.second_shift.set("None")
        self.third_day.set("None")
        self.third_shift.set("None")
        
    def run(self):
        """Starts the GUI application"""
        self.root.mainloop()

if __name__ == "__main__":
    app = EmployeeSchedulerGUI()
    app.run()